[Overlay] Fix dock widget floating behavior and general rework/stability issues.

MisterMakerNL opened the pull request:
After trying a 1000 things I think it works pretty good now!
It removes all overlay title bar stuff, which in my opinion isn't necessary for floating windows.. 
This fixes the issue that the window was moving down after every edit.

<!-- Include a brief summary of the changes. -->

<!--
The FreeCAD community thanks you for your contribution!
By creating a Pull Request you agree to the contributing policy. The complete policy can be found in the root of the source tree (CONTRIBUTING.md) or at https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md

This template provides guidance on creating a PR that can be reviewed and approved as quickly as possible. Comments may be safely deleted.

Unless you know exactly what you're doing, please leave the checkbox 'Allow edits by maintainers' enabled.  This will allow maintainers to help you.
-->

## Issues
<!-- link to individual issues this PR closes by referencing the issue number (e.g., fixes #1234, closes #4321). -->
fixes https://github.com/FreeCAD/FreeCAD/issues/14650

## Before and After Images
<!-- If your proposed changes affect the FreeCAD GUI, add before and after screenshots -->
after
<img width="3840" height="2055" alt="image" src="https://github.com/user-attachments/assets/ee7c4caa-115d-4afb-94ba-f10e0b321ef2" />



<!--  Notes on the PR Review Process

The following section describes what the maintainers consider when reviewing your Pull Request.  These items may not require you to take any action.  This information is provided for context. Understanding what we consider will help you prepare your request for speedy approval.

You can find additional documentation about these guidelines in the [Developers handbook](https://freecad.github.io/DevelopersHandbook).

Alignment (Does the PR align with the goals and interests of the project?)
  - Does the PR have at least one issue linked, which this PR closes?
  - Has the conversation on the PR and related issue(s) reached consensus?
  - If the PR affects the GUI, is the Design Working Group (DWG) aware and have they had time to review and comment?
  - If the PR affects the GUI, did the contributor include before/after images?
  - If the PR affects standards and workflow, is the CAD Working Group (CWG) aware and have they had time to review/comment?

Impact (Does the change affect other parts of the project?)
  - Has the impact on documentation been considered and appropriate action taken?
  - Has the impact on translation been considered appropriate action taken?
  - Will the PR affect existing user documents?

Code Quality (Is code well-written and maintainable?)
  - Does the PR warrant a review by the Code Quality Working Group (CQWG)?
  - Does the change include tests?
  - Is the PR rebased on the current main branch with unnecessary commits squashed?

Release (Are there considerations related to release timing?)
  - Has the PR been considered for backporting to the latest release branch?
  - Have the release notes been considered/updated?
  -->

MisterMakerNL committed:
Fix dock widget floating behavior and title bar handling

Improves handling of QDockWidget floating state by dynamically setting or removing the custom title bar widget based on the floating status. Also fixes a bug in OverlayTitleBar::endDrag() where the wrong widget was being hidden when dragging ended. Ensures dock widgets are properly removed from all overlays when floated.

MisterMakerNL committed:
Update OverlayWidgets.cpp

MisterMakerNL committed:
Works awesome

Ok this gone a bit overboard, and I really lost trach what has changed but it works awesome.

MisterMakerNL commented:
Ok so that title bar thing was kind of a big deal...
Now I kinda let Copilot go all lose on this.
Normally I wouldn't want it to do stuff I don't understand but man this Overlay code was a really complex.
So yeah, this time it not just kinda works, it works. And it works good! Even overlay still works on undocked window.
https://github.com/user-attachments/assets/d6e6e024-aa7d-45da-83b6-a27b7b546263

MisterMakerNL committed:
Show dock title in minimal overlay title bar

In minimal mode, the overlay title bar now displays the dock's window title, ensuring better context and usability for floating or frameless docks. The title is elided if necessary and drawn centered, improving visual clarity.

MisterMakerNL committed:
Add resize support for floating dock titlebars

Implemented mouse-based resizing for floating QDockWidgets by detecting edge drags in OverlayTitleBar. Added logic to update the cursor when hovering near edges and to handle resize start, update, and finish events. Introduced new private members and an enum in OverlayWidgets.h to track resize state and edge.

MisterMakerNL committed:
Update OverlayManager.cpp

Had some issues with missing titlebar when switching overlay on and off. Now everything seems to run smooth.

MisterMakerNL commented:
Had one minor issue where the title bar would dissapear, this was fixed by undocking and docking the window and now works very smooth.

xtemp09 commented:
I don't have building environment for Windows, so I cannot test or fix something. I have to say that this commit blows up my mind. I just looked at lines 1583 and below. This is insanity! Does it work?


@MisterMakerNL, I think it would be better for everyone if the code made sense. If the code is compilable, it does not mean it is readable.


@MisterMakerNL, are you sure that the code does exactly what you say in the PR description?

MisterMakerNL commented:
So to explain what is going on here.
The overlay replaces the titlebars. It has a titlebar for docked and undocked.
This was done by some custom code.
Somewhere in that code caused the window to walk.
Now Qt has an option to use a minimal titlebar which is basically a second titlebar with your custom titlebar. I replaced the custom system with this system.
Now I got some issues with dragging the window, and getting a ghost window and jagged lines, so I replaced that.
Then the undocked window was a pain too resize with the mouse so added that. Which appeared to be a lot of code.
And last I had issues with the titlebar disappearing when the docked window toggled overlay. Fixed that by docking and undocking the window.
Now everything seems to run fin, even better and the walking window is fixed.

xtemp09 commented:
I'm just saying that the code is human-unreadable.

MisterMakerNL committed:
Refactor OverlayManager API for clarity and consistency

Renamed several OverlayManager methods for improved clarity (e.g., setupTitleBar -> setupOverlayTitleBar, initDockWidget -> initializeDockForOverlay, unsetupDockWidget -> cleanupDockWidget, actionsForDock -> getActionsForDock, setMouseTransparent -> setMousePassthroughEnabled, refresh -> refreshOverlays). Added backwards-compatible wrappers for deprecated methods to maintain API stability. Updated all usages in related files to use the new method names.

MisterMakerNL committed:
Update OverlayManager.cpp

Sorry for adding more code, but the titlebar didn't update when the docked window was moved from left to right etc.

MisterMakerNL committed:
Update OverlayWidgets.cpp

Sorry more bugs, more fixes.

MisterMakerNL committed:
Update OverlayManager.cpp

Still the title bar didn't stick.... Now it does.

MisterMakerNL committed:
Comment out setupTitleBar calls and adjust tab position

Disabled multiple calls to setupTitleBar in OverlayManager, likely for debugging or to prevent custom title bar setup. Also changed the tab position for the right dock area in OverlayTabWidget from East to West.

MisterMakerNL committed:
Removed useless titlebar.

Removed useless titlebar.

MisterMakerNL committed:
Delete show_taskwatcher_macro.FCMacro

woops

MisterMakerNL committed:
Update OverlayManager.cpp

Titlebars whent missing again...

MisterMakerNL committed:
added extra code to reset the panels when overlay got switched off.

MisterMakerNL committed:
Use native window decorations for floating docks

Removes FramelessWindowHint and WA_TranslucentBackground attributes from floating QDockWidget overlays to avoid compositor and translucency bugs. Native window decorations are now used for improved stability and compatibility.

MisterMakerNL committed:
Improve dark mode detection for OverlayTabWidget

Enhanced isStyleSheetDark to detect dark mode based on OS-specific methods: Windows registry, macOS environment variable, and Linux theme variables. This provides more accurate dark mode detection across platforms instead of relying on stylesheet string matching.

MisterMakerNL committed:
Improve dark theme detection in isStyleSheetDark

Update isStyleSheetDark to prioritize Spreadsheet.TextColor preference when a stylesheet is set, avoiding OS dark mode checks in that case. This ensures theme detection aligns with user preferences in the Spreadsheet module.

MisterMakerNL committed:
Refactor dark theme detection and logging in GUI modules

Replaces string-based stylesheet checks with palette-based dark theme detection in OverlayWidgets and TechDrawGuiUtil. Adds platform-specific logic for dark mode detection in DrawGuiUtil, and updates related function signatures and usages. Improves logging in OverlayManager and adds copyright/license header to OverlayWidgets.h.

MisterMakerNL committed:
Use StylesheetIconsColor for dark theme detection

Replaces Spreadsheet.TextColor preference with StylesheetIconsColor in theme preferences to determine if a stylesheet uses a dark or light theme. The value is now case-insensitive and defaults to light theme if unrecognized or missing.

MisterMakerNL committed:
Merge branch 'FreeCAD:main' into Fix_walking_window

MisterMakerNL committed:
Update OverlayWidgets.h

small fix

MisterMakerNL committed:
Update title bar background fill logic

Refines the background fill behavior in OverlayTitleBar::paintEvent. Now avoids filling the background if the palette's Window color is transparent, making floating titlebars less obtrusive and improving appearance when stylesheets do not apply.

MisterMakerNL committed:
icon naming fixes

MisterMakerNL committed:
Icon fixes.

I fixed the icons, they should work now with no-stylesheet with native OS.
They also do not need to be set by the stylesheet anymore.

maxwxyz commented:
@MisterMakerNL there is a conflict and CI fails

MisterMakerNL committed:
trying to fix issue

MisterMakerNL committed:
Update OverlayWidgets.cpp

trying to fix github issue
